### PR TITLE
Add Documented and annotations to builder methods

### LINKED ISSUE #20: Add propagation of properties' annotations
Hello, (it's me again). I've made a PR (https://github.com/skinny85/jilt/pull/19) with the additional feature to propagate properties' annotations in the builders. I think this would be nice to have (in fact, I'd like to use it for my projects). Open to discussion/rejection, obviously!

### PR DESCRIPTION
**Pull request by thmasker**
Fixes #20

### CODE REVIEW

File: src/main/java/org/jilt/internal/AbstractBuilderGenerator.java [hunk 1]
```diff
@@ -331,12 +331,18 @@ protected final TypeVariableName mangleTypeParameter(TypeVariableName typeVariab
     }
 
     private ParameterSpec setterParameterSpec(VariableElement attribute, TypeName parameterType) {
-        ParameterSpec.Builder ret = ParameterSpec.builder(parameterType,
-                this.attributeSimpleName(attribute));
-        AnnotationMirror nullableAnnotation = this.firstAnnotationCalledNullable(attribute);
+        ParameterSpec.Builder ret = ParameterSpec.builder(parameterType, this.attributeSimpleName(attribute));
+        AnnotationMirror nullableAnnotation = firstAnnotationCalledNullable(attribute);
```
Comment (skinny85) [inline src/main/java/org/jilt/internal/AbstractBuilderGenerator.java:335]: Can you please revert these changes? They are functionally the same, so they're just making the diff in the PR bigger.

```suggestion
        ParameterSpec.Builder ret = ParameterSpec.builder(parameterType,
                this.attributeSimpleName(attribute));
        AnnotationMirror nullableAnnotation = this.firstAnnotationCalledNullable(attribute);
```

File: src/main/java/org/jilt/internal/AbstractBuilderGenerator.java [hunk 2]
```diff
@@ -331,12 +331,18 @@ protected final TypeVariableName mangleTypeParameter(TypeVariableName typeVariab
...
         if (nullableAnnotation != null) {
             ret.addAnnotation(AnnotationSpec.get(nullableAnnotation));
         }
+
+        for (AnnotationMirror annotation : attribute.getAnnotationMirrors()) {
+            if (!annotation.equals(nullableAnnotation)) {
+                ret.addAnnotation(AnnotationSpec.get(annotation));
+            }
+        }
```
Comment (skinny85) [inline src/main/java/org/jilt/internal/AbstractBuilderGenerator.java:344]: It doesn't really make much sense to copy `@Nullable` annotations separately if we're copying all annotations anyway.

Let's unify the two parts:

```suggestion
        for (AnnotationMirror annotation : attribute.getAnnotationMirrors()) {
            ret.addAnnotation(AnnotationSpec.get(annotation));
        }
```

File: src/main/java/org/jilt/internal/AbstractBuilderGenerator.java [hunk 3]
```diff
@@ -331,12 +331,18 @@ protected final TypeVariableName mangleTypeParameter(TypeVariableName typeVariab
...
+
```
Comment (skinny85) [inline src/main/java/org/jilt/internal/AbstractBuilderGenerator.java:345]: I feel like we're copying too much here - what if the annotation is from a field (because `@Builder` was placed on the class level) , but it's not allowed to be placed on a parameter? We would be generating code that doesn't compile here.

Can you please add a check that makes sure the given annotation can be placed on a parameter before you copy it?

File: src/test/java/org/jilt/test/PropagatedAnnotationsTest.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
+package org.jilt.test;
```
Comment (skinny85) [inline src/test/java/org/jilt/test/PropagatedAnnotationsTest.java:1]: Please indent this file with 4 spaces, not 2, like the rest of the project.

File: src/test/java/org/jilt/test/data/annotations/PropagatedAnnotationName.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
+package org.jilt.test.data.annotations;
```
Comment (skinny85) [inline src/test/java/org/jilt/test/data/annotations/PropagatedAnnotationName.java:1]: Same here (please indent this file with 4 spaces, not 2, like the rest of the project).

File: src/test/java/org/jilt/test/data/annotations/PropagatedAnnotationName.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+import javax.annotation.Nonnull;
+
+@Builder(style = BuilderStyle.STAGED_PRESERVING_ORDER)
+public class PropagatedAnnotationName {
+  public final String firstName;
+  public final String middleName;
+
+  @Nonnull
+  @CheckForNull
+  public final String lastName;
```
Comment (skinny85) [inline src/test/java/org/jilt/test/data/annotations/PropagatedAnnotationName.java:16]: Also annotate this field with an annotation that cannot be placed on a parameter, to make sure the filtering works correctly.

If you can't find an annotation like this in the current project or its dependencies, feel free to define one next to this class.

**Review by skinny85 — COMMENTED** [review]
Overall looks great, thanks a lot for opening the PR!

Just a few minor comments before we merge this one in.

File: src/main/java/org/jilt/internal/AbstractBuilderGenerator.java (hunk 1 above)
Comment (thmasker) [reply src/main/java/org/jilt/internal/AbstractBuilderGenerator.java:335]: Sorry about this!

File: src/main/java/org/jilt/internal/AbstractBuilderGenerator.java (hunk 2 above)
Comment (thmasker) [reply src/main/java/org/jilt/internal/AbstractBuilderGenerator.java:345]: You are absolutely right about this

File: src/test/java/org/jilt/test/data/annotations/NonParameterAnnotation.java [hunk 7]
```diff
@@ -0,0 +1,9 @@
+package org.jilt.test.data.annotations;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Target;
+
+@Target({ElementType.TYPE, ElementType.FIELD, ElementType.METHOD, ElementType.CONSTRUCTOR, ElementType.LOCAL_VARIABLE,
+        ElementType.ANNOTATION_TYPE, ElementType.PACKAGE, ElementType.TYPE_PARAMETER})
```
Comment (thmasker) [inline src/test/java/org/jilt/test/data/annotations/NonParameterAnnotation.java:7]: I did not include `MODULE` and `RECORD_COMPONENT` to prevent compatibility checks from failing

**Comment by skinny85** [comment]
> Sorry for the mess with the commits. Backwards compatibility is wonderful in this library, but a mess when coding haha

Don't worry about it. I'll squash-merge the PR into one commit anyway.

File: src/main/java/org/jilt/internal/AbstractBuilderGenerator.java [hunk 8]
```diff
@@ -333,13 +334,19 @@ protected final TypeVariableName mangleTypeParameter(TypeVariableName typeVariab
     private ParameterSpec setterParameterSpec(VariableElement attribute, TypeName parameterType) {
         ParameterSpec.Builder ret = ParameterSpec.builder(parameterType,
                 this.attributeSimpleName(attribute));
-        AnnotationMirror nullableAnnotation = this.firstAnnotationCalledNullable(attribute);
-        if (nullableAnnotation != null) {
-            ret.addAnnotation(AnnotationSpec.get(nullableAnnotation));
-        }
+        addAnnotations(ret, attribute.getAnnotationMirrors());
+        addAnnotations(ret, attribute.asType().getAnnotationMirrors());
```
Comment (skinny85) [inline src/main/java/org/jilt/internal/AbstractBuilderGenerator.java:338]: Let's prefix the calls with `this`:

```suggestion
        this.addAnnotations(ret, attribute.getAnnotationMirrors());
        this.addAnnotations(ret, attribute.asType().getAnnotationMirrors());
```

File: src/main/java/org/jilt/internal/AbstractBuilderGenerator.java [hunk 9]
```diff
@@ -333,13 +334,19 @@ protected final TypeVariableName mangleTypeParameter(TypeVariableName typeVariab
...
     }
 
+    private void addAnnotations(final ParameterSpec.Builder ret, final List<? extends AnnotationMirror> annotationMirrors) {
+        for (final AnnotationMirror annotation : annotationMirrors) {
+            if (AnnotationUtils.isAnnotationAllowedInParameter(annotation)) {
+                ret.addAnnotation(AnnotationSpec.get(annotation));
+            }
+        }
+    }
+
```
Comment (skinny85) [inline src/main/java/org/jilt/internal/AbstractBuilderGenerator.java:349]: No `final` on local variables or method parameters:

```suggestion
    private void addAnnotations(ParameterSpec.Builder ret, List<? extends AnnotationMirror> annotationMirrors) {
        for (AnnotationMirror annotation : annotationMirrors) {
            if (AnnotationUtils.isAnnotationAllowedInParameter(annotation)) {
                ret.addAnnotation(AnnotationSpec.get(annotation));
            }
        }
    }
```

File: src/main/java/org/jilt/internal/AbstractBuilderGenerator.java [hunk 10]
```diff
@@ -333,13 +334,19 @@ protected final TypeVariableName mangleTypeParameter(TypeVariableName typeVariab
...
-        AnnotationMirror nullableAnnotation = this.firstAnnotationCalledNullable(attribute);
-        if (nullableAnnotation != null) {
-            ret.addAnnotation(AnnotationSpec.get(nullableAnnotation));
-        }
+        addAnnotations(ret, attribute.getAnnotationMirrors());
+        addAnnotations(ret, attribute.asType().getAnnotationMirrors());
         return ret.build();
     }
 
+    private void addAnnotations(final ParameterSpec.Builder ret, final List<? extends AnnotationMirror> annotationMirrors) {
```
Comment (skinny85) [inline src/main/java/org/jilt/internal/AbstractBuilderGenerator.java:342]: This method name could be better. `addAnnotationsToParam()` maybe?

File: src/main/java/org/jilt/utils/AnnotationUtils.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    static {
+        try {
+            NON_PARAMETER_ANNOTATIONS.add((ElementType) ElementType.class.getField("RECORD_COMPONENT").get(null));
+            NON_PARAMETER_ANNOTATIONS.add((ElementType) ElementType.class.getField("MODULE").get(null));
+        } catch (NoSuchFieldException e) {
+            // Java version does not support RECORD_COMPONENT
+        } catch (IllegalAccessException e) {
+            // Java version does not support RECORD_COMPONENT
+        }
+    }
```
[the remark this answers is not in the document]
Comment (skinny85) [reply src/main/java/org/jilt/utils/AnnotationUtils.java:23]: I actually don't think this is the correct way to check whether an annotation can be placed on a parameter.

An annotation can be placed on a parameter if either it's not annotated with `@Target` at all (which means it can be applied to any element), or, if `@Target` is present, `@Target.value` contains either `ElementType.PARAMETER` or `ElementType.TYPE_USE`. All other `ElementType` values are immaterial.

So, you can safely get rid of `NON_PARAMETER_ANNOTATIONS`, and everything related to it.

Given how simple that check is, please get rid of `AnnotationUtils`, and move the logic into `AbstractBuilderGenerator`.

File: src/test/java/org/jilt/test/data/annotations/NonParameterAnnotation.java (hunk 7 above)
Comment (skinny85) [reply src/test/java/org/jilt/test/data/annotations/NonParameterAnnotation.java:7]: Like I said above, this is immaterial. Change it to have just `ElementType.FIELD`.

File: src/main/java/org/jilt/internal/AbstractBuilderGenerator.java (hunk 1 above)
Comment (skinny85) [reply src/main/java/org/jilt/internal/AbstractBuilderGenerator.java]: No worries 🙂. Thanks for the revert!

File: src/main/java/org/jilt/utils/AnnotationUtils.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (thmasker) [reply src/main/java/org/jilt/utils/AnnotationUtils.java]: I felt I was overcomplicating thing, and indeed I was! Thanks for your comments

File: src/main/java/org/jilt/internal/AbstractBuilderGenerator.java [hunk 12]
```diff
@@ -333,13 +335,34 @@ protected final TypeVariableName mangleTypeParameter(TypeVariableName typeVariab
...
-        AnnotationMirror nullableAnnotation = this.firstAnnotationCalledNullable(attribute);
-        if (nullableAnnotation != null) {
-            ret.addAnnotation(AnnotationSpec.get(nullableAnnotation));
-        }
+        this.addAnnotationsToParam(ret, attribute.getAnnotationMirrors());
+        this.addAnnotationsToParam(ret, attribute.asType().getAnnotationMirrors());
         return ret.build();
     }
 
+    private void addAnnotationsToParam(ParameterSpec.Builder ret, List<? extends AnnotationMirror> annotationMirrors) {
```
Comment (skinny85) [inline src/main/java/org/jilt/internal/AbstractBuilderGenerator.java:343]: Let's just call the first parameter `param`, and the second parameter `annotations`:

```suggestion
    private void addAnnotationsToParam(ParameterSpec.Builder param, List<? extends AnnotationMirror> annotations) {
```

File: src/main/java/org/jilt/internal/AbstractBuilderGenerator.java [hunk 13]
```diff
@@ -333,13 +335,34 @@ protected final TypeVariableName mangleTypeParameter(TypeVariableName typeVariab
...
+        for (AnnotationMirror annotation : annotationMirrors) {
+            if (this.isAnnotationAllowedInParameter(annotation)) {
+                ret.addAnnotation(AnnotationSpec.get(annotation));
+            }
+        }
+    }
+
+    public boolean isAnnotationAllowedInParameter(AnnotationMirror annotation) {
```
Comment (skinny85) [inline src/main/java/org/jilt/internal/AbstractBuilderGenerator.java:351]: There are a few small issues with this method:

1. It shouldn't be `public`.
2. I think the correct word here is "on", not "in".
3. Let's use "param" instead of "parameter", to be consistent with the name above.

Putting that all together, this should be:

```suggestion
    private boolean isAnnotationAllowedOnParam(AnnotationMirror annotation) {
```

File: src/main/java/org/jilt/internal/AbstractBuilderGenerator.java [hunk 14]
```diff
@@ -333,13 +335,34 @@ protected final TypeVariableName mangleTypeParameter(TypeVariableName typeVariab
...
+        Target targetAnnotation = annotation.getAnnotationType().asElement().getAnnotation(Target.class);
+        if (targetAnnotation == null) {
+            return true;
+        }
+
+        for (ElementType elementType : targetAnnotation.value()) {
+            if (ElementType.TYPE_USE.equals(elementType) || ElementType.PARAMETER.equals(elementType)) {
```
Comment (skinny85) [inline src/main/java/org/jilt/internal/AbstractBuilderGenerator.java:358]: Let's flip the order in this check, `PARAMETER` is the more important one:

```suggestion
            if (ElementType.PARAMETER.equals(elementType) || ElementType.TYPE_USE.equals(elementType)) {
```

Comment (skinny85) [inline src/main/java/org/jilt/internal/AbstractBuilderGenerator.java:358]: `ElementType.TYPE_USE` is actually only available since Java 8, and this library still supports Java 6.

Compare it against a string instead:

```suggestion
            if (ElementType.PARAMETER.equals(elementType) || "TYPE_USE".equals(elementType.name())) {
```

File: src/test/java/org/jilt/test/PropagatedAnnotationsTest.java [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+        assertThat(firstNameSetter.getParameters()[0].getDeclaredAnnotation(CheckForNull.class)).isNotNull();
+
+        Method middleNameSetter = PropagatedAnnotationNameBuilder.class.getMethod("middleName", String.class);
+
+        assertThat(middleNameSetter.getParameters()[0].getDeclaredAnnotation(Nullable.class)).isNotNull();
+
+        Method lastNameSetter = PropagatedAnnotationNameBuilder.class.getMethod("lastName", String.class);
+
+        assertThat(lastNameSetter.getParameters()[0].getDeclaredAnnotation(Nonnull.class)).isNotNull();
+        assertThat(lastNameSetter.getParameters()[0].getDeclaredAnnotation(NonParameterAnnotation.class)).isNull();
```
Comment (skinny85) [inline src/test/java/org/jilt/test/PropagatedAnnotationsTest.java:29]: Let's group these together - this is pretty difficult to read right now:

```suggestion
        Method firstNameSetter = PropagatedAnnotationNameBuilder.class.getMethod("firstName", String.class);
        assertThat(firstNameSetter.getParameters()[0].getDeclaredAnnotation(CheckForNull.class)).isNotNull();

        Method middleNameSetter = PropagatedAnnotationNameBuilder.class.getMethod("middleName", String.class);
        assertThat(middleNameSetter.getParameters()[0].getDeclaredAnnotation(Nullable.class)).isNotNull();

        Method lastNameSetter = PropagatedAnnotationNameBuilder.class.getMethod("lastName", String.class);
        assertThat(lastNameSetter.getParameters()[0].getDeclaredAnnotation(Nonnull.class)).isNotNull();
        assertThat(lastNameSetter.getParameters()[0].getDeclaredAnnotation(NonParameterAnnotation.class)).isNull();
```

File: src/test/java/org/jilt/test/PropagatedAnnotationsTest.java [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+        assertThat(firstNameSetter.getParameters()[0].getDeclaredAnnotation(CheckForNull.class)).isNotNull();
+
+        Method middleNameSetter = PropagatedAnnotationNameBuilders.MiddleName.class.getMethod("middleName", String.class);
+
+        assertThat(middleNameSetter.getParameters()[0].getDeclaredAnnotation(Nullable.class)).isNotNull();
+
+        Method lastNameSetter = PropagatedAnnotationNameBuilders.LastName.class.getMethod("lastName", String.class);
+
+        assertThat(lastNameSetter.getParameters()[0].getDeclaredAnnotation(Nonnull.class)).isNotNull();
+        assertThat(lastNameSetter.getParameters()[0].getDeclaredAnnotation(NonParameterAnnotation.class)).isNull();
```
Comment (skinny85) [inline src/test/java/org/jilt/test/PropagatedAnnotationsTest.java:45]: Same here:

```suggestion
        Method firstNameSetter = PropagatedAnnotationNameBuilders.FirstName.class.getMethod("firstName", String.class);
        assertThat(firstNameSetter.getParameters()[0].getDeclaredAnnotation(CheckForNull.class)).isNotNull();

        Method middleNameSetter = PropagatedAnnotationNameBuilders.MiddleName.class.getMethod("middleName", String.class);
        assertThat(middleNameSetter.getParameters()[0].getDeclaredAnnotation(Nullable.class)).isNotNull();

        Method lastNameSetter = PropagatedAnnotationNameBuilders.LastName.class.getMethod("lastName", String.class);
        assertThat(lastNameSetter.getParameters()[0].getDeclaredAnnotation(Nonnull.class)).isNotNull();
        assertThat(lastNameSetter.getParameters()[0].getDeclaredAnnotation(NonParameterAnnotation.class)).isNull();
```

File: src/test/java/org/jilt/test/data/annotations/PropagatedAnnotationName.java [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+
+import org.jilt.Builder;
+import org.jilt.BuilderStyle;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+@Builder(style = BuilderStyle.STAGED_PRESERVING_ORDER)
+public class PropagatedAnnotationName {
```
Comment (skinny85) [inline src/test/java/org/jilt/test/data/annotations/PropagatedAnnotationName.java:11]: Maybe call this class `PropagatedAnnotationsFullName` instead?

```suggestion
public class PropagatedAnnotationsFullName {
```

File: src/test/java/org/jilt/test/data/annotations/NonParameterAnnotation.java [hunk 18]
```diff
@@ -0,0 +1,8 @@
+package org.jilt.test.data.annotations;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Target;
+
+@Target({ElementType.FIELD})
```
Comment (skinny85) [inline src/test/java/org/jilt/test/data/annotations/NonParameterAnnotation.java:6]: I think you can drop the `{}` with one element?

```suggestion
@Target(ElementType.FIELD)
```

**Review by skinny85 — COMMENTED** [review]
Looks great! We're down to the last nitpick comments 🙂.

File: src/main/java/org/jilt/internal/AbstractBuilderGenerator.java (hunk 13 above)
Comment (thmasker) [reply src/main/java/org/jilt/internal/AbstractBuilderGenerator.java:351]: copy pasting is the devil... haha. I like your thoroughness!

**Review by skinny85 — APPROVED** [review]
Looks perfect, thanks @thmasker!
